File: C++/maxNumber.cpp
/**
 * @File    : maxNumber.cpp
 * @Brief   : 拼接最大数
 * @Link    : https://leetcode-cn.com/problems/create-maximum-number/
 * @Author  : Wei Li
 * @Date    : 2021-08-25
*/

#include <iostream>
#include <vector>
#include <stack>
#include <algorithm>

class Solution
{
public:
    std::vector<int> maxNumber(std::vector<int> &nums1, std::vector<int> &nums2, int k)
    {
        int m = nums1.size(), n = nums2.size();
        std::vector<int> maxSubsequence(k, 0);
        int start = std::max(0, k - n), end = std::min(k, m);
        for (int i = start; i <= end; i++)
        {
            std::vector<int> subsequence1(MaxSubsequence(nums1, i));
            std::vector<int> subsequence2(MaxSubsequence(nums2, k - i));
            std::vector<int> curMaxSubsequence(merge(subsequence1, subsequence2));
            if (compare(curMaxSubsequence, 0, maxSubsequence, 0) > 0)
            {
                maxSubsequence.swap(curMaxSubsequence);
            }
        }
        return maxSubsequence;
    }

    std::vector<int> MaxSubsequence(std::vector<int> &nums, int k)
    {
        int length = nums.size();
        std::vector<int> stack(k, 0);
        int top = -1;
        int remain = length - k;
        for (int i = 0; i < length; i++)
        {
            int num = nums[i];
            while (top >= 0 && stack[top] < num && remain > 0)
            {
                top--;
                remain--;
            }
            if (top < k - 1)
            {
                stack[++top] = num;
            }
            else
            {
                remain--;
            }
        }
        return stack;
    }

    std::vector<int> merge(std::vector<int> &subsequence1, std::vector<int> &subsequence2)
    {
        int x = subsequence1.size(), y = subsequence2.size();
        if (x == 0)
        {
            return subsequence2;
        }
        if (y == 0)
        {
            return subsequence1;
        }
        int mergeLength = x + y;
        std::vector<int> merged(mergeLength);
        int index1 = 0, index2 = 0;
        for (int i = 0; i < mergeLength; i++)
        {
            if (compare(subsequence1, index1, subsequence2, index2) > 0)
            {
                merged[i] = subsequence1[index1++];
            }
            else
            {
                merged[i] = subsequence2[index2++];
            }
        }
        return merged;
    }

    int compare(std::vector<int> &subsequence1, int index1, std::vector<int> &subsequence2, int index2)
    {
        int x = subsequence1.size(), y = subsequence2.size();
        while (index1 < x && index2 < y)
        {
            int difference = subsequence1[index1] - subsequence2[index2];
            if (difference != 0)
            {
                return difference;
            }
            index1++;
            index2++;
        }
        return (x - index1) - (y - index2);
    }
};

template <typename T>
void DisplayElement(const T &container)
{
    for (auto element : container)
    {
        std::cout << element << " ";
    }
    std::cout << std::endl;
}

// -------------------------------
int main(int argc, char **argv)
{
    std::vector<int> nums1{3, 9};
    std::vector<int> nums2{8, 9};
    int k = 3;
    // std::vector<int> nums1{3, 4, 6, 5};
    // std::vector<int> nums2{9, 1, 2, 5, 8, 3};
    // int k = 5;

    Solution solution;
    auto max_number_order = solution.maxNumber(nums1, nums2, k);
    DisplayElement(max_number_order);

    return 0;
}

File: Python/maxNumber.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: maxNumber.py
@Function: 拼接最大数
@Link: https://leetcode-cn.com/problems/create-maximum-number/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-08-25
"""

from typing import List


class Solution:
    def maxNumber(self, nums1: List[int], nums2: List[int], k: int) -> List[int]:
        m, n = len(nums1), len(nums2)
        maxSubsequence = [0] * k
        start, end = max(0, k - n), min(k, m)

        for i in range(start, end + 1):
            subsequence1 = self.getMaxSubsequence(nums1, i)
            subsequence2 = self.getMaxSubsequence(nums2, k - i)
            curMaxSubsequence = self.merge(subsequence1, subsequence2)
            if self.compare(curMaxSubsequence, 0, maxSubsequence, 0) > 0:
                maxSubsequence = curMaxSubsequence

        return maxSubsequence

    def getMaxSubsequence(self, nums: List[int], k: int) -> int:
        stack = [0] * k
        top = -1
        remain = len(nums) - k

        for i, num in enumerate(nums):
            while top >= 0 and stack[top] < num and remain > 0:
                top -= 1
                remain -= 1
            if top < k - 1:
                top += 1
                stack[top] = num
            else:
                remain -= 1

        return stack

    def merge(self, subsequence1: List[int], subsequence2: List[int]) -> List[int]:
        x, y = len(subsequence1), len(subsequence2)
        if x == 0:
            return subsequence2
        if y == 0:
            return subsequence1

        mergeLength = x + y
        merged = list()
        index1 = index2 = 0

        for _ in range(mergeLength):
            if self.compare(subsequence1, index1, subsequence2, index2) > 0:
                merged.append(subsequence1[index1])
                index1 += 1
            else:
                merged.append(subsequence2[index2])
                index2 += 1

        return merged

    def compare(self, subsequence1: List[int], index1: int, subsequence2: List[int], index2: int) -> int:
        x, y = len(subsequence1), len(subsequence2)
        while index1 < x and index2 < y:
            difference = subsequence1[index1] - subsequence2[index2]
            if difference != 0:
                return difference
            index1 += 1
            index2 += 1

        return (x - index1) - (y - index2)


# -------------------------
if __name__ == "__main__":
    # nums1 = [6, 7]
    # nums2 = [6, 0, 4]
    # k = 5

    nums1 = [3, 4, 6, 5]
    nums2 = [9, 1, 2, 5, 8, 3]
    k = 5

    solution = Solution()
    max_number_order = solution.maxNumber(nums1, nums2, k)
    print(f"The solution of this problem is {max_number_order}")
